xarray/core/duck_array_ops.py

```diff
@@ -373,7 +373,7 @@ def sum_where(data, axis=None, dtype=None, where=None):
 def where(condition, x, y):
     """Three argument where() with better dtype promotion rules."""
     xp = get_array_namespace(condition, x, y)
-    return xp.where(condition, *as_shared_dtype([x, y], xp=xp))
+    return xp.where(xp.astype(condition, xp.bool), *as_shared_dtype([x, y], xp=xp))
```

Comment:
looks like it is not that simple... at some point we're trying to call
```python
duck_array_ops.where(True, np.arange(5), np.float64(np.nan))
```
where `numpy` implements `astype` by forwarding to the `ndarray` method (and our own wrapper does the same).

To resolve this, we probably need to make `cond` a 0d array:
```suggestion
    try:
        cond_ = astype(condition, xp.bool)
    except AttributeError:
        cond_ = xp.asarray(condition, dtype=xp.bool)
    return xp.where(cond_, *as_shared_dtype([x, y], xp=xp))
```
I'm not sure I like the implications of this.

Reply:
Hi @keewis ! I got stuck on the same thing. What I did:

```python
# Check array type
    if isinstance(condition, np.ndarray):
        condition = condition.astype(dtype=np.bool)

    elif isinstance(condition, xp_strict._array_object.Array):
        condition = xp_strict.asarray(condition, dtype=xp_strict.bool)
    else:
        raise TypeError("Unsupported array type")

    xp = get_array_namespace(condition, x, y)
    return xp.where(condition, *as_shared_dtype([x, y], xp=xp))
```
This passes `test_array_api` but `test_computation` fails with a bunch of `TypeErrors`

Reply:
thanks, that gave me some ideas on how to possibly fix this in a cleaner way (and I also found a potential bug in the groupby code)